### PR TITLE
Use swift-driver triple

### PR DESCRIPTION
**Pull request by rauhul**
Removes and replaces the custom LLVM Triple type from the Basics module with the implementation from SwiftDriver. This change allows SwiftPM to understand all triples that swift-driver already does and resolves some issues surrounding incorrect triple parsing. For example the Basics.Triple was not able to parse 'armv7em-apple-none-eabihf-macho' where as SwiftDriver.Triple can.

### CODE REVIEW

File: Sources/PackageModel/UserToolchain.swift [hunk 1]
```diff
@@ -382,11 +382,11 @@ public final class UserToolchain: Toolchain {
                             // the merged swiftmodules.  XCTest followed suit.
                             "-I",
                             AbsolutePath(
-                                validating: "usr/lib/swift/windows/\(triple.arch)",
+                                validating: "usr/lib/swift/windows/\(triple.arch?.rawValue ?? "unknown")",
```
Comment (compnerd) [inline Sources/PackageModel/UserToolchain.swift:385]: I'm somewhat uncomfortable with this.  The `unknown` here is confusing.  Would we be able to replace this with something like `<<NULL>>` or something that is explicitly bad and obviously so?

File: Tests/BasicsTests/TripleTests.swift [hunk 2]
```diff
@@ -76,9 +76,51 @@ final class TripleTests: XCTestCase {
         XCTAssertTriple("i686-pc-windows-gnu", isApple: false, isDarwin: false)
         XCTAssertTriple("i686-pc-windows-cygnus", isApple: false, isDarwin: false)
     }
-
+
```
Comment (compnerd) [inline Tests/BasicsTests/TripleTests.swift:79]: Nit: bleeding whitespace

**Comment by MaxDesiatov** [comment]
Is there a reason we're not using Driver's `Triple` type instead of copying it here?

**Comment by rauhul** [comment]
@swift-ci please smoke test

**Comment by MaxDesiatov** [comment]
I've previously copied this `Triple` implementation from TSC since we wanted to deprecate the TSC version, but it seems like the duplication in both TSC and Driver is historic. As SwiftPM already uses Driver as a dependency, I don't see why we can't use Driver's `Triple` type directly.

**Comment by MaxDesiatov** [comment]
@swift-ci test windows

File: Tests/BasicsTests/TripleTests.swift [hunk 3]
```diff
@@ -81,4 +81,46 @@ final class TripleTests: XCTestCase {
...
+        XCTAssertTriple("i686-apple-macos10.12.0", forPlatformVersion: "", is: "i686-apple-macos")
+        XCTAssertTriple("i686-apple-macos10.12.0", forPlatformVersion: "13.0", is: "i686-apple-macos13.0")
+
+        XCTAssertTriple("riscv64-apple-darwin", forPlatformVersion: "", is: "riscv64-apple-darwin")
+        XCTAssertTriple("riscv64-apple-darwin", forPlatformVersion: "22", is: "riscv64-apple-darwin22")
+
+        XCTAssertTriple("mips-apple-darwin19", forPlatformVersion: "", is: "mips-apple-darwin")
+        XCTAssertTriple("mips-apple-darwin19", forPlatformVersion: "22", is: "mips-apple-darwin22")
+    }
```
Comment (MaxDesiatov) [inline Tests/BasicsTests/TripleTests.swift:125]: Could we also test for something like `armv7em-apple-none-eabihf-macho` you mentioned in the PR description, or for whatever other triple we'd like to support with more than three components?

Comment (rauhul) [reply Tests/BasicsTests/TripleTests.swift:125]: done!

**Comment by rauhul** [comment]
@swift-ci please smoke test

**Comment by rauhul** [comment]
@swift-ci please smoke test windows

File: Sources/Basics/Triple+Basics.swift [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+//
+// Copyright (c) 2014-2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import enum TSCBasic.JSON
```
Comment (MaxDesiatov) [inline Sources/Basics/Triple+Basics.swift:13]: One last request, but otherwise this LGTM: can we have a comment in either of these files or in README.md of this directory that these files are vendored from Swift Driver, maybe also with an explanation why? Without it I don't think we make it obvious why the directory is called `Vendor` in the first place without opening this PR and reading this discussion to get the context.

Comment (rauhul) [reply Sources/Basics/Triple+Basics.swift:13]: Yeah for sure, I'll push that diff later tonight/tomorrow

File: Sources/PackageModel/UserToolchain.swift (hunk 1 above)
Comment (tomerd) [reply Sources/PackageModel/UserToolchain.swift]: should this throw in case it is null?

File: Sources/PackageModel/UserToolchain.swift [hunk 5]
```diff
@@ -716,32 +716,32 @@ public final class UserToolchain: Toolchain {
                 // (~5.7), we always had a singular installed SDK.  Prefer the
                 // new variant which has an architecture subdirectory in `bin`
                 // if available.
-                switch triple.arch {
-                case .x86_64, .x86_64h:
+                switch triple.archName {
```
[the remark this answers is not in the document]
Comment (tomerd) [reply Sources/PackageModel/UserToolchain.swift]: switching on the string feels like a regression from a type safety pov. is it necessary?

File: Sources/SPMBuildCore/BinaryTarget+Extensions.swift [hunk 6]
```diff
@@ -43,8 +43,8 @@ extension BinaryTarget {
         // Filter the libraries that are relevant to the triple.
         // FIXME: this filter needs to become more sophisticated
         guard let library = metadata.libraries.first(where: {
-            $0.platform == triple.os.asXCFrameworkPlatformString &&
-                $0.architectures.contains(triple.arch.rawValue)
+            $0.platform == triple.os?.asXCFrameworkPlatformString &&
+                $0.architectures.contains(triple.archName)
```
Comment (tomerd) [inline Sources/SPMBuildCore/BinaryTarget+Extensions.swift:47]: can architectures be made more type safe instead of string array?

File: Sources/Basics/Triple.swift [hunk 7]
```diff
@@ -10,173 +10,36 @@
...
-    fileprivate static func parseVersion(_ string: String) -> String? {
-        let candidate = String(string.drop(while: { $0.isLetter }))
-        if candidate != string && !candidate.isEmpty {
-            return candidate
-        }
-
-        return nil
-    }
+extension Triple {
+    public static let macOS = try! Self("x86_64-apple-macosx")
```
Comment (tomerd) [inline Sources/Basics/Triple.swift]: does Triple need a non-throwing constructer that takes a static string (can precondition)

File: Sources/Basics/Triple.swift [hunk 8]
```diff
@@ -10,173 +10,36 @@
...
-        self.abiVersion = abiVersion
-    }
-
-    fileprivate static func parseOS(_ string: String) -> OS? {
-        var candidates = OS.allCases.map { (name: $0.rawValue, value: $0) }
-        // LLVM target triples support this alternate spelling as well.
-        candidates.append((name: "macos", value: .macOS))
-        return candidates.first(where: { string.hasPrefix($0.name) })?.value
+extension Triple {
+    public init(_ description: String) throws {
```
Comment (tomerd) [inline Sources/Basics/Triple.swift]: where is this used? worried about the public extension

File: Sources/Basics/Triple.swift [hunk 9]
```diff
@@ -10,173 +10,36 @@
 //
 //===----------------------------------------------------------------------===//
 
-import protocol Foundation.CustomNSError
-import var Foundation.NSLocalizedDescriptionKey
-
 import enum TSCBasic.JSON
 import class TSCBasic.Process
+import struct SwiftDriver.Triple
```
Comment (tomerd) [inline Sources/Basics/Triple.swift]: This file should probably be renamed to Triple+Extension to match style of repo

Comment (tomerd) [inline Sources/Basics/Triple.swift]: seems like some of these imports can be removed?

File: Package.swift [hunk 10]
```diff
@@ -152,6 +152,7 @@ let package = Package(
                 .product(name: "DequeModule", package: "swift-collections"),
                 .product(name: "OrderedCollections", package: "swift-collections"),
                 .product(name: "SwiftToolsSupport-auto", package: "swift-tools-support-core"),
+                .product(name: "SwiftDriver", package: "swift-driver"),
```
Comment (tomerd) [inline Package.swift]: hmm I dont love adding this to Basics. I think we are going that only for the extension which are a bit questionable imho. Maybe we can limit the dependency to the modules that actually use it and find a better home (closer to call site) for the extensions

**Comment by WowbaggersLiquidLunch** [comment]
> Is there a reason we're not using Driver's `Triple` type directly instead of copying it here?

> We don't want to introduce a dependency on Driver in the "data model" portion of SwiftPM since that will cause issues for the Xcode integration.

What if `Triple` is abstracted into its own module?

**Comment by tomerd** [comment]
I think creating a new library that has some of these very basic models (eg semantic version, triple) is called for. TSC was suppose to be that but became a grab bag of "stuff" that is impossible to version properly and as such lost utility and will be deprecated. while we are working to deprecate TSC, this specific solution seems like an okay compromise

File: Sources/PackageModel/UserToolchain.swift (hunk 1 above)
Comment (compnerd) [reply Sources/PackageModel/UserToolchain.swift]: I think that having it throw would be a much safer option.  We need to know the architecture to be able to build.

File: Sources/SPMBuildCore/BinaryTarget+Extensions.swift (hunk 6 above)
Comment (rauhul) [reply Sources/SPMBuildCore/BinaryTarget+Extensions.swift:47]: I have absolutely no idea how this XCFramework code works and I'd like to avoid touching it if possible in this PR.
```
struct XCFrameworkMetadata.Library {
    let libraryIdentifier: String
    let libraryPath: String
    let headersPath: String?
    let platform: String
    let architectures: [String]
}
```

File: Sources/Basics/Triple.swift (hunk 9 above)
Comment (rauhul) [reply Sources/Basics/Triple.swift]: I think this is stale

File: Sources/PackageModel/UserToolchain.swift (hunk 1 above)
Comment (rauhul) [reply Sources/PackageModel/UserToolchain.swift]: Resolved this by using `archName` which is non-optional

File: Sources/PackageModel/UserToolchain.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (rauhul) [reply Sources/PackageModel/UserToolchain.swift]: Changed this back to switch on arch, the coverage of the included cases is increased as a result though.

File: Sources/Basics/Triple.swift (hunk 7 above)
Comment (rauhul) [reply Sources/Basics/Triple.swift]: Probably, but I want to avoid additional changes API changes in this PR. (The existing triple uses the above API shape)

File: Sources/Basics/Triple.swift (hunk 8 above)
Comment (rauhul) [reply Sources/Basics/Triple.swift]: I added this to match the current Basics.Triple API

File: Sources/Basics/Triple.swift (hunk 9 above)
Comment (rauhul) [reply Sources/Basics/Triple.swift]: Fixed!

**Comment by tomerd** [comment]
@rauhul is this ready to be merged?

**Comment by rauhul** [comment]
@tomerd I just pushed fixes to the review feedback, do you taking another look?

File: Sources/Basics/Triple+Basics.swift (hunk 4 above)
Comment (rauhul) [reply Sources/Basics/Triple+Basics.swift:13]: done!

**Comment by rauhul** [comment]
Likely in a follow on PR `Sources/Basics/Triple+Basics.swift` should be slimmed down as much as possible, upstreaming necessary content to swift-driver

**Comment by rauhul** [comment]
@swift-ci please smoke test

**Comment by rauhul** [comment]
@swift-ci please smoke test windows

**Comment by rauhul** [comment]
@swift-ci please test windows

**Comment by rauhul** [comment]
@swift-ci please smoke test

**Comment by rauhul** [comment]
@swift-ci please test windows

**Comment by rauhul** [comment]
@swift-ci test Windows

**Comment by rauhul** [comment]
@swift-ci test windows
